src/FSharp.Core/prim-types.fs

```diff
@@ -4069,6 +4072,28 @@ namespace Microsoft.FSharp.Core
 
     and 'T voption = ValueOption<'T>
 
+#if !NET8_0_OR_GREATER
+namespace System.Runtime.CompilerServices
+    open System
+    open System.ComponentModel
+    open Microsoft.FSharp.Core
+
+    [<CompilerMessage("This type is for compiler use and should not be used directly", 1204, IsHidden=true)>]
+    [<Sealed>]
+    [<AttributeUsage(AttributeTargets.Class ||| AttributeTargets.Struct ||| AttributeTargets.Interface, Inherited = false)>]
+    type CollectionBuilderAttribute (builderType: Type, methodName: string) =
+        inherit Attribute ()
+        member _.BuilderType = builderType
+        member _.MethodName = methodName
+
+    [<CompilerMessage("This type is for compiler use and should not be used directly", 1204, IsHidden=true)>]
+    [<EditorBrowsable(EditorBrowsableState.Never)>]
+    [<Sealed>]
+    [<AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)>]
+    type ScopedRefAttribute () =
+        inherit Attribute ()
+#endif
```

Comment:
So, this will always be true, since fslib is ns20 and ns21. Won't this be an issue when loading it in net9 project?

Reply:
> Won't this be an issue when loading it in net9 project?

These polyfills should actually be `internal`. Once I push that, I don't think that should be an issue.

You're right that we could technically remove the ifdef here — the main reason to keep it would be for it to serve as a reminder to remove the polyfill if we ever added a higher target framework.

Reply:
Right, but at runtime, won't it be the problem -  there will be a polyfil and one from runtime. I remember us having some issues with IsReadOnlyAttribute (I think)

Reply:
> I remember us having some issues with IsReadOnlyAttribute (but we generated it - #15313).

Does that mean that this should be fine, since the compiler doesn't (and, I think, won't) generate this attribute? It seems to work fine in the C# tests I added.

Reply:
We should probably test for it, at least manually

Reply:
I'll add some tests.

Reply:
This is nice.

Reply:
I replaced the ifdef with a comment in 94a8541 (even though there is still [another, similar ifdef](https://github.com/dotnet/fsharp/blob/43bcb29ca5dd0958db7636133d92b60eb61cd9a6/src/FSharp.Core/prim-types.fs#L384-L385) in the same file).

From my understanding of #15313 and from Kevin's comment https://github.com/dotnet/fsharp/pull/17359#discussion_r1658115159, I can't think of a way that defining these attributes internally and putting them on `List<'T>` and `Set<'T>` will cause a problem for consumers.